src/grammar.coffee

```diff
@@ -468,6 +469,112 @@ grammar =
+  ]
+
+  JsxElementShorthands: [
+    o 'JsxElementShorthandClasses',                                                  -> classes: $1
+    o 'JsxElementShorthandId',                                                       -> id: $1
+    o 'JsxElementShorthandId JsxElementShorthandClasses',                            -> classes: $2, id: $1
+    o 'JsxElementShorthandClasses JsxElementShorthandId JsxElementShorthandClasses', -> classes: $1.concat($2), id: $2
+    o 'JsxElementShorthandClasses JsxElementShorthandId',                            -> classes: $1, id: $2
+  ]
+
+  JsxElementShorthandId: [
+    o 'JSX_ID_SHORTHAND_SYMBOL JSX_ID_SHORTHAND', -> $2
+  ]
+
+  JsxElementShorthandClasses: [
+    o 'JsxElementShorthandClass',                            -> [$1]
+    o 'JsxElementShorthandClasses JsxElementShorthandClass', -> $1.concat $2
+  ]
+
+  JsxElementShorthandClass: [
+    o 'JSX_CLASS_SHORTHAND_SYMBOL JSX_CLASS_SHORTHAND', -> $2
+  ]
+
+  JsxAttributes: [
+    o 'JsxParenthesizedAttributes_',                      -> list: $1
+    o 'JsxParenthesizedAttributes_ JsxObjectAttributes_', -> list: $1, object: $2
+    o 'JsxObjectAttributes_ JsxParenthesizedAttributes_', -> list: $2, object: $1
+    o 'JsxObjectAttributes_',                             -> object: $1
+  ]
+
+  JsxParenthesizedAttributes_: [
+    o 'JSX_PARENTHESIZED_ATTRIBUTES_START JsxParenthesizedAttributes JSX_PARENTHESIZED_ATTRIBUTES_END', -> $2
+  ]
+
+  JsxParenthesizedAttributes: [
+    o 'JsxParenthesizedAttribute',                            -> [$1]
+    o 'JsxParenthesizedAttributes JsxParenthesizedAttribute', -> $1.concat $2
+    o 'INDENT JsxParenthesizedAttributes OUTDENT',            -> $2
+    o 'JsxParenthesizedAttributes TERMINATOR',                -> $1
+    o 'JsxParenthesizedAttributes INDENT JsxParenthesizedAttributes OUTDENT', -> $1.concat $3
+  ]
+
+  JsxParenthesizedAttribute: [
+    o 'JSX_ATTRIBUTE_NAME = JsxParenthesizedAttributeValue', -> name: $1, value: $3
+  ]
+
+  JsxParenthesizedAttributeValue: [
+    o 'STRING',         -> new StringLiteral $1
+    o '{ Expression }', -> $2
+  ]
+
+  JsxObjectAttributes_: [
+    o 'JSX_OBJECT_ATTRIBUTES_START JsxAttributesObject JSX_OBJECT_ATTRIBUTES_END', -> $2
+  ]
+
+  JsxAttributesObject: [
+    o '{ AssignList OptComma }', -> new JsxAttributesObj $2, $1.generated
+  ]
+
+  JsxElementChildren_: [
+    o 'JsxElementChildren JSX_ELEMENT_INLINE_BODY_END', -> $1
+    o 'INDENT JsxElementChildren OUTDENT',              -> $2
+  ]
+
+  JsxElementChildren: [
+    o 'JsxElementChild',                    -> [$1]
+    o 'JsxElementChildren JsxElementChild', -> $1.concat $2
+  ]
+
+  JsxElementChild: [
+    o 'JSX_ELEMENT_CONTENT',           -> $1
+    o 'JsxElement',                    -> $1
+    o '{ Expression }',                -> $2
+    o '{ INDENT Expression OUTDENT }', -> $3
+  ]
```

Comment:
There’s over 100 lines of new grammar rules here. This would be a significant burden to maintain.